### ISSUE TITLE
Wrong overloaded method selected when compiling with error-prone

### ISSUE THREAD

**Issue by ePaul** [google/error-prone#507]
When calling an overloaded varargs method from another varargs method using its varargs parameter), sometimes not the most specific one is selected.

I'm using the maven setup as described in [the documentation](http://errorprone.info/docs/installation), [full project here](https://github.com/ePaul/error-prone-compiler-bug). (That project has a second example with Hamcrest where even a wrongly typed matcher is returned.)

```java
package example;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

public class ExampleTest {

    @SafeVarargs
    final <E> String example(Class<? super E>... classes) {
        return "with classes";
    }

    @SafeVarargs
    final <E> String example(E... items) {
        return "with items";
    }

    @SafeVarargs
    final String indirectCall(Class<? super ExampleTest>... classes) {
        return example(classes);
    }

    @Test
    public void testNormalVarargsCall() {
        final String result = example(ExampleTest.class, Object.class);
        assertEquals("with classes", result);
    }

    @Test
    public void testIndirectVarargsCall() {
        final String result = indirectCall(ExampleTest.class, Object.class);
        assertEquals("with classes", result);
    }
}
```
With a stock Java 8 compiler (1.8.0_45) both tests are green.
When compiling with the "javac-with-errorprone" as indicated by your documentation (i.e `mvn test -P error-prone` in my example project), the second test is failing (the first one is still green).

(If you make both overloaded methods even have incompatible return types, you'll get a compiler error.)

I guess this is a bug in your frozen upstream Java 9 javac. I didn't yet try to compile it with a real Javac 9 to see whether it is already fixed there (I have no Java 9 set up here).

**Comment by cushon** [comment]
That's [JDK-8159680](https://bugs.openjdk.java.net/browse/JDK-8159680), which was fixed in jdk-9+124. We can include the backport in the next error-prone-javac release.

**Comment by ePaul** [comment]
That would be good. (It doesn't seem like exactly the same issue on the first look, as that one results in a compiler error, and my one in choosing the wrong method. Though when removing the `example(E... items)` method, I get an error instead.)

I guess it is not possible to simply use the standard compiler of the JDK where we are running instead of having to maintain a fork with backports?

**Comment by eaftan** [comment]
@ePaul No, we depend on javac internals that change from version to version, so we have to fix our version of javac for Error Prone to work correctly.  Our javac version also contains upstream fixes for bugs in the released javac as well.

**Comment by cushon** [comment]
Backported as google/error-prone-javac@a94f7d8af1bede00976366e472cc7d2124d75ec2

**Comment by cushon** [comment]
The fix has been released in 2.0.16.
